docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1163 @@
+> enough" can be estimated based on the timing assumption in this
+> specification. 
+
+
+#### **[NewFS-A-STATUS-INTERVAL]**:
+Let Sigma be the (upper bound on the)
+time between two calls of `QueryStatus()`.
+
+#### **[NewFS-A-GOOD-PERIOD]**:
+
+A time interval *[begin,end]* is *good period* if:
+
+- *fmax* is the number of faulty peers in *peerIDs* at time *begin*
+- *end >= begin + 2 Delta (fmax + 3)*
+- no faulty peer is added before time *end*
+
+> In the analysis below we assume that the termination condition of
+> *Fastsync* V2 [FS-V2-TIMEOUT] (that is, termination when the height is not incremented for
+> some time), is replaced by the termination condition
+> *height = TargetHeight* in the postcondition of
+> `Execute`. Therefore, [NewFS-A-STATUS-INTERVAL] does not interfere
+> with in analysis. If a correct peer reports a new height "shortly
+> before termination" this leads to an additional round trip to
+> request and add the new block. Then [NewFS-A-DELTA] ensures that
+> *Fastsync* catches up.
+
+Arguments: 
+
+1. If a faulty peer *p* reports a faulty block, `SequentialVerify` will
+  eventually remove *p* from *peerIDs*
+  
+2. By `SequentialVerify`, if a faulty peer *p* reports multiple faulty
+  blocks, *p* will be removed upon trying to check the block with the
+  smallest height received from *p*.
+
+3. Assume whenever a block does not have an open request, `CreateRequest` is
+   called immediately, which calls `Block(n)` on a peer. Say this
+   happens at time *t*. There are two cases: 
+  
+   - by t + 2 Delta a block is added to *blockStore*
+   - at t + 2 Delta `Block(n)` timed out and *n* is removed from
+       peer.
+	   
+4. Let *f(t)* be the number of faulty peers in *peerIDs* at time *t*;  
+   *f(begin) = fmax*.
+	   
+5. Let t_i be the sequence of times `OnBlockResponse(addr,b)` is
+   invoked or times out with *b.Height = height + 1*.
+   
+6. By 3., 
+   - (a). *t_1 <= begin + 2 Delta* 
+   - (b). *t_{i+1} <= t_i + 2 Delta* 
+
+7. By an inductive argument we prove for *i > 0* that
+
+   - (a). *height(t_{i+1}) > height(t_i)*, or
+   - (b). *f(t_{i+1}) < f(t_i))* and *height(t_{i+1}) = height(t_i)*  
+
+   Argument: if the peer is faulty and does not return a block, the
+   peer is removed, if it is faulty and returns a faulty block
+   `SequentialVerify` removes the peer (b). If the returned block is OK,
+   height is increased (a).
+	 
+8. By 2. and 7., faulty peers can delay incrementing the height at
+   most *fmax* times, where each time "costs" *2 Delta* seconds. We
+   have additional *2 Delta* initial offset (3a) plus *2 Delta* to get
+   all missing blocks after the last fault showed itself. (This
+   assumes that an arbitrary number of blocks can be obtained and
+   checked within one round-trip 2 Delta; which either needs
+   conservative estimation of Delta, or a more refined analysis). Thus
+   we reach the *targetHeight* and terminate by time *end*.
```

Comment:
I like this analysis. This was my intuition, but I wasn't able to formalise it. I am just wondering if faulty peers sending statusResponse for nonexistent blocks can prevent us from termination. I am wondering also if the fact that correct peers are not fully up to date with the blockchain matters. It seems that what this proposal ensures is that once we hit max height of correct peers, if faulty peers advertise higher height they will eventually be removed as *2 Delta < ETIME*.